### PR TITLE
Allow locals, parameters, range variables inside a lambda to shadow names outside the lambda

### PR DESCRIPTION
**Pull request by cston**

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/LocalFunctionTests.cs [hunk 1]
```diff
@@ -4481,91 +4481,6 @@ static void M()
             comp.VerifyDiagnostics();
         }
 
-        [Fact]
-        public void ShadowNames_LambdaInsideLocalFunction_01()
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/LocalFunctionTests.cs]: These two tests were moved to LambdaTests.cs.

**Comment by cston** [comment]
@jaredpar, @agocke, @dotnet/roslyn-compiler, please review.

File: src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs [hunk 2]
```diff
@@ -2941,5 +2941,253 @@ void method1()
...
+                Diagnostic(ErrorCode.ERR_QueryRangeVariableOverrides, "x").WithArguments("x").WithLocation(14, 38));
+
+            comp = CreateCompilation(source, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics();
+
+            comp = CreateCompilation(source);
+            comp.VerifyDiagnostics();
+        }
+
+        // PROTOTYPE: Need similar tests for queries as for lambdas.
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs]: PROTOTYPE comment

File: src/Compilers/CSharp/Portable/Binder/Binder.cs [hunk 3]
```diff
@@ -201,6 +201,11 @@ internal virtual bool IsLabelsScopeBinder
             }
         }
 
+        /// <summary>
+        /// True if this is the top-level binder for a local function or lambda.
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Binder/Binder.cs]: Is it worth adding clarification for how logical lambdas, like LINQ, operate here?

File: src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs [hunk 4]
```diff
@@ -2941,5 +2941,289 @@ void method1()
...
+                // (17,21): error CS0136: A local or parameter named 'x' cannot be declared in this scope because that name is used in an enclosing local scope to define a local or parameter
+                //                 int x = 0;
+                Diagnostic(ErrorCode.ERR_LocalIllegallyOverrides, "x").WithArguments("x").WithLocation(17, 21),
+                // (18,39): error CS0412: 'T': a parameter, local variable, or local function cannot have the same name as a method type parameter
+                //                 Action<int> b2 = (int T) => { };
+                Diagnostic(ErrorCode.ERR_LocalSameNameAsTypeParam, "T").WithArguments("T").WithLocation(18, 39));
+
+            comp = CreateCompilation(source);
+            comp.VerifyDiagnostics();
+        }
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs]: Shouldn't we have testing for anonymous functions (aka the `delegate` syntax)?

**Comment by jaredpar** [comment]
Done with review pass (iteration 4). Question about anonymous functions.

File: src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs (hunk 4 above)
Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs]: Yes, thanks.

---
In reply to: [251062142](https://github.com/dotnet/roslyn/pull/32754#discussion_r251062142) [](ancestors = 251062142)

File: src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs [hunk 5]
```diff
@@ -113,7 +113,7 @@ internal static class ParameterHelpers
             var typeParameters = (object)methodOwner != null ?
                 methodOwner.TypeParameters :
                 default(ImmutableArray<TypeParameterSymbol>);
-            bool allowShadowingNames = binder.Compilation.IsFeatureEnabled(MessageID.IDS_FeatureStaticLocalFunctions) &&
+            bool allowShadowingNames = binder.Compilation.IsFeatureEnabled(MessageID.IDS_FeatureNameShadowingInNestedFunctions) &&
                 methodOwner?.MethodKind == MethodKind.LocalFunction;
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs]: Is this not used for lambdas?

File: src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs [hunk 6]
```diff
@@ -2941,5 +2941,401 @@ void method1()
...
```
Comment (agocke) [inline src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs]: Do we have a test for hiding with nesting? e.g. `Task<int, Task<int, int>> t = x => x => x;`?

Also maybe a test that `_` is still a variable, not a discard? So `Task<int, Task<int, int>> t = _ => _ => _;` works but `Task<int, int, int> t = (_, _) => 0` fails?

File: src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs (hunk 5 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs]: Added an assert.

---
In reply to: [251087156](https://github.com/dotnet/roslyn/pull/32754#discussion_r251087156) [](ancestors = 251087156,251082666)

**Pull request merged**